Evaluate typeof window at compile time

jsnajdr opened the pull request:
A little webpack optimization that should help us eliminate some browser-only code from the server bundle and vice versa. Evaluate `typeof window` at compile time, to be either `"undefined"` or `"object"`. Next.js uses this optimization, too:

https://github.com/vercel/next.js/blob/5117cc6ea990a6ddda5585a5fe2bf6a1914285c8/packages/next/src/build/babel/loader/get-config.ts#L105

I got the idea when working on #84243 and wondering why the server bundle contains some modules for Redux middlewares that should be loaded only in browser:

https://github.com/Automattic/wp-calypso/blob/dfa89637b1b84d9f21f223cb3fc26d664b8793a6/client/state/index.ts#L39-L49

jsnajdr committed:
Evaluate typeof window at compile time

tyxla commented on packages/calypso-build/webpack.config.js:
```diff
@@ -125,6 +125,7 @@ function getWebpackConfig(
 		node: false,
 		plugins: [
 			new webpack.DefinePlugin( {
+				'typeof window': JSON.stringify( 'object' ),
```
Same as above:

```suggestion
				'typeof window': 'object',
```

tyxla commented on client/webpack.config.node.js:
```diff
@@ -159,6 +159,7 @@ const webpackConfig = {
 			} ),
 		new webpack.ExternalsPlugin( 'commonjs', getExternals() ),
 		new webpack.DefinePlugin( {
+			'typeof window': JSON.stringify( 'undefined' ),
```
Is there a good reason to use `JSON.stringify()` on a string literal? I'd just go with the literal:

```suggestion
			'typeof window': 'undefined',
```

jsnajdr replied:
I think we need to make it a string with quotes: `'"undefined"'`. That's what `JSON.stringify` is doing. It's true that we can easily achieve the goal even without it, but the surrounding code already uses it anyway.

tyxla replied:
Ah, got it! Thanks, let's move forward then.

tyxla reviewed (approved):
That's a smart one 👍 Thanks!

jsnajdr committed:
Optimize also the main browser config

jsnajdr commented:
I also added the optimization to the main `client/webpack.config.js`. Previously I added it only to the server config (`.node.js`) and the `calypso-build` config, but the browser config for Calypso is not based on the `calypso-build` one.

Now ICFY is showing a little decrease in the bundle size. That's expected, as the code guarded by `typeof window === 'undefined' is eliminated, and do are the checks themselves.
